packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts

```diff
@@ -40,8 +64,109 @@ export const useGridInfiniteLoader = (apiRef: GridApiRef): void => {
       apiRef.current.publishEvent(GRID_ROWS_SCROLL_END, rowScrollEndParam);
       isInScrollBottomArea.current = true;
     }
-  }, [options, containerSizes, apiRef, visibleColumns]);
+  }, [logger, options, containerSizes, apiRef, visibleColumns]);
+
+  const handleGridVirtualPageChange = React.useCallback(
+    (params: GridVirtualPageChangeParams) => {
+      logger.debug('Virtual page changed');
+      if (!containerSizes || !options.getRows) {
+        return;
+      }
+
+      let nextPage = params.nextPage;
+
+      if (
+        params.nextPage > params.currentPage &&
+        params.nextPage !== renderState.renderedSizes!.lastPage
+      ) {
+        nextPage = params.nextPage + 1;
+      }
+
+      const newRowsBatchStartIndex = nextPage * containerSizes.viewportPageSize;
+      const toBeLoadedRange: any = [...allRows].splice(
+        newRowsBatchStartIndex,
+        containerSizes.viewportPageSize,
+      );
+
+      if (!toBeLoadedRange.includes(null)) {
+        return;
+      }
+
+      const { rows }: GridGetRowsReturnValue = options.getRows({
+        startIndex: newRowsBatchStartIndex,
+        viewportPageSize: containerSizes.viewportPageSize,
+        sortModel,
+        filterModel: filterState,
+        api: apiRef,
+      });
+
+      if (rows.length) {
+        apiRef.current.loadRows(newRowsBatchStartIndex, containerSizes.viewportPageSize, rows);
+      }
+    },
+    [logger, options, renderState, allRows, sortModel, containerSizes, filterState, apiRef],
+  );
+
+  const handleGridSortModelChange = React.useCallback(
+    (params: GridSortModelParams) => {
+      logger.debug('Sort model changed');
+
+      if (!containerSizes || !options.getRows) {
+        return;
+      }
+
+      const newRowsBatchStartIndex = renderState.virtualPage * containerSizes.viewportPageSize;
+      const { rows, rowCount }: GridGetRowsReturnValue = options.getRows({
+        startIndex: newRowsBatchStartIndex,
+        viewportPageSize: containerSizes.viewportPageSize,
+        sortModel: params.sortModel,
+        filterModel: filterState,
+        api: apiRef,
+      });
+
+      if (rows.length) {
+        apiRef.current.loadRows(
+          newRowsBatchStartIndex,
+          containerSizes.viewportPageSize,
+          rows,
+          rowCount,
+        );
+      }
+    },
+    [logger, options, renderState, containerSizes, filterState, apiRef],
+  );
+
+  const handleGridFilterModelChange = React.useCallback(
+    (params: GridFilterModelParams) => {
+      logger.debug('Filter model changed');
+      if (!containerSizes || !options.getRows) {
+        return;
+      }
+
+      const newRowsBatchStartIndex = renderState.virtualPage * containerSizes.viewportPageSize;
+      const { rows, rowCount }: GridGetRowsReturnValue = options.getRows({
```

Comment:
Shouldn't we expect a Promise? IMO, the use case for infinite loading is when the full set of rows is not available _a priori_.

Reply:
we never call options.handler directly, we use the events

Reply:
Updated